### ISSUE TITLE
Package:web usage for migration

### ISSUE THREAD

**Issue by ebala96** [dart-lang/web#229]
I'm doing a migration to 'package:web' and 'dart:interop'. I'm currently using the context object from the 'dart:js' package to store functions in the window object reference. However, I've noticed that there's no direct way to access the context from the 'package:web'.

I need help with this migration. Below is the code snippet I need to migrate

```
void _connectJsToFlutter({VoidCallback? then}) {
    js.context['$jsToDartConnectorFN$iframeViewType'] = (js.JsObject window) {
      jsWindowObject = window;
      for (final cb in widget.dartCallBacks) {
        jsWindowObject[cb.name] = cb.callBack;
      }

      jsWindowObject[webOnClickInsideIframeCallback] = (onClickCallbackObject) {
        _handleOnIframeClick(onClickCallbackObject as String);
      };

      webViewXController.connector = jsWindowObject;

      then?.call();
    };
}
```

Can you provide the correct usage of context and JsObject needed for the migration?

**Comment by srujzs** [comment]
`globalContext` is the `dart:js_interop` equivalent. Also note that we use static interfaces instead of class wrappers now with `dart:js_interop`.

A simple migration might look like the following:

```dart
import 'dart:js_interop';
import 'dart:js_interop'; // to dynamically set properties using []=
...
void _connectJsToFlutter({VoidCallback? then}) {
    globalContext['$jsToDartConnectorFN$iframeViewType'] = (JSObject window) {
      jsWindowObject = window;
      for (final cb in widget.dartCallBacks) {
        jsWindowObject[cb.name] = cb.callBack.toJS;
      }

      jsWindowObject[webOnClickInsideIframeCallback] = (String onClickCallbackObject) {
        _handleOnIframeClick(onClickCallbackObject);
      }.toJS;

      webViewXController.connector = jsWindowObject;

      then?.call();
    };
}
```

It's a little difficult to tell what the types of some of these objects are, but a few key points here are:
- you need to use either a primitive type or an interop type when making a call to JS
- functions are passed using `.toJS`, which will return a new `JSFunction` that will call the original function. These functions can also only take in/return primitive types and interop types

For more information, look at https://dart.dev/interop/js-interop/usage and https://dart.dev/interop/js-interop/js-types

**Comment by ebala96** [comment]
Thank you. Understood the usage now.

Also could you let me know if I am doing the right way of doing onbeforeunload.

```
window.onbeforeunload = () async{
      //my code snippet
}.toJS;
```

**Comment by srujzs** [comment]
Yup, both should work. You can also use `EventStreamProvider` in `lib/src/helpers/events/streams.dart` if you'd rather work with `Stream`s: `EventStreamProvider<BeforeUnloadEvent>('beforeunload').forTarget(window)`.

**Comment by ebala96** [comment]
Also I am doing callJsMethod and passing `List<dynamic> params` as the args parameter. How can I do a List<dynamic> to JSArray?

**Comment by srujzs** [comment]
It's better practice to define an interface for the `JSObject` using an extension type than use `dart:js_interop_unsafe` (see https://dart.dev/interop/js-interop/usage#interop-type-members on how to write an interop extension type), but if you don't statically know the method name, then `dart:js_interop_unsafe` and `callMethodVarArgs` is ok as the replacement.

> How can I pass the List as [List<JSAny?>? arguments]?

What does the `List` contain? You'll need to convert the contents of the `List` to `JSAny`s. If it's a `List<String>` for example, you can do a `list.map((e) => e.toDart).toList()` to get a `List<JSString>`. Or if you have a literal e.g. `['hello', 'world']`, you can just do `['hello'.toJS, 'world'.toJS]` instead.

**Comment by ebala96** [comment]
List contains dynamic type values.
It contains string, boolean, number and Set/List.

**Comment by srujzs** [comment]
Assuming the nested `Set`s/`List`s all contain types that can be converted as well (primitives or other nested structures), one thing you could try is using [`jsify`](https://api.dart.dev/main/6a25f12f40463d74eea9de79deef5bb1a4b0f763/dart-js_interop/NullableObjectUtilExtension/jsify.html) which will convert all the entries in that list recursively.

```dart
List<dynamic> list = ...;
List<JSAny?> jsList = list.map((e) => jsify(e)).toList();
```

**Comment by ebala96** [comment]
Where can I get more information or examples of JS interop usage other than https://dart.dev/interop/js-interop/usage?

**Comment by srujzs** [comment]
Besides the `js-interop` page and its subsections, I'd look at some examples of existing migrations like those in https://github.com/flutter/packages (just grep for `dart:js_interop`) or https://github.com/flutter/devtools. We have some tests in the Dart SDK as well that might be useful: https://github.com/dart-lang/sdk/tree/main/tests/lib/js/static_interop_test.

Of course, if you feel like you still have some questions, feel free to ask. If you want to provide feedback on what documentation might be useful for you or others, you can do so in https://github.com/dart-lang/site-www. There's some ongoing work there to get a "getting started" tutorial.
